apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_subs.erl

```diff
@@ -85,91 +104,126 @@
-                    _ = emqx_external_broker:add_persistent_route(TopicFilter, SessionId),
-                    {SubId, S1} = emqx_persistent_session_ds_state:new_id(S0),
-                    {SStateId, S2} = emqx_persistent_session_ds_state:new_id(S1),
-                    SState = #{
-                        parent_subscription => SubId, upgrade_qos => UpgradeQoS, subopts => SubOpts
-                    },
-                    S3 = emqx_persistent_session_ds_state:put_subscription_state(
-                        SStateId, SState, S2
-                    ),
-                    Subscription = #{
-                        id => SubId,
-                        current_state => SStateId,
-                        start_time => now_ms()
-                    },
-                    S = emqx_persistent_session_ds_state:put_subscription(
-                        TopicFilter, Subscription, S3
-                    ),
-                    ?tp(persistent_session_ds_subscription_added, #{
-                        topic_filter => TopicFilter, session => SessionId
-                    }),
-                    {ok, S};
-                false ->
-                    {error, ?RC_QUOTA_EXCEEDED}
-            end;
-        Sub0 = #{current_state := SStateId0, id := SubId} ->
-            SState = #{parent_subscription => SubId, upgrade_qos => UpgradeQoS, subopts => SubOpts},
-            case emqx_persistent_session_ds_state:get_subscription_state(SStateId0, S0) of
-                SState ->
-                    %% Client resubscribed with the same parameters:
-                    {ok, S0};
-                _ ->
-                    %% Subsription parameters changed:
-                    {SStateId, S1} = emqx_persistent_session_ds_state:new_id(S0),
-                    S2 = emqx_persistent_session_ds_state:put_subscription_state(
-                        SStateId, SState, S1
-                    ),
-                    Sub = Sub0#{current_state => SStateId},
-                    S = emqx_persistent_session_ds_state:put_subscription(TopicFilter, Sub, S2),
-                    {ok, S}
-            end
+        %% This is a new subscription:
+        undefined when NSubscriptions < MaxSubscriptions ->
+            S = create_subscription(SessionId, TopicFilter, SubOpts, Props, S0),
+            {ok, S};
+        undefined when NSubscriptions >= MaxSubscriptions ->
+            {error, ?RC_QUOTA_EXCEEDED};
+        %% This is existing subscription:
+        Sub = #{} ->
+            Mode = get_subscription_mode(Sub),
+            case subscription_mode(TopicFilter, SubOpts, Props) of
+                Mode ->
+                    %% ...And the mode hasn't changed, update it:
+                    S = update_subscription(TopicFilter, SubOpts, Sub, Props, S0);
+                _Different ->
+                    %% ...But the mode is different, recreate it:
+                    S1 = delete_subscription(SessionId, TopicFilter, Sub, S0),
```

Comment:
With `SubId` changed, how scheduler's `update_stream_subscription_state_ids` will work? Seems it won't be able to update stream's sub states 🤔 Will they become orphaned?

Reply:
~~I believe they will be taken care of by `emqx_persistent_session_ds_subs:gc/1` eventually.~~ Good catch, scheduler needs to be informed about deleted subscriptions.

Reply:
If we create a new subscription, it will obtain new streams (starting from a new point?). This will probably lead (unlike `update_subscription` case) to the loss of continuity of message flow.

Reply:
Yes. This is currently considered fine, since we expect `upgrade_qos` is rarely used and thus even rarer toggled in runtime. Please share if you think this needs to be reevaluated.